statsmodels/tsa/bds.py

```diff
@@ -0,0 +1,242 @@
+    if epsilon is None:
+        epsilon = distance * x.std(ddof=1)
+
+    return np.abs(x[:, None] - x) < epsilon
+
+
+def correlation_sum(indicators, embedding_dim):
+    """
+    Calculate a correlation sum
+
+    Useful as an estimator of a correlation integral
+
+    Parameters
+    ----------
+    indicators : 2d array
+        matrix of distance threshold indicators
+    embedding_dim : integer
+        embedding dimension
+
+    Returns
+    -------
+    corrsum : float
+        Correlation sum
+    indicators_joint
+        matrix of joint-distance-threshold indicators
+
+    """
+    if not indicators.ndim == 2:
+        raise ValueError('Indicators must be a matrix')
+    if not indicators.shape[0] == indicators.shape[1]:
+        raise ValueError('Indicator matrix must be symmetric (square)')
+
+    if embedding_dim == 1:
+        indicators_joint = indicators
+    else:
+        corrsum, indicators = correlation_sum(indicators, embedding_dim - 1)
+        indicators_joint = indicators[1:, 1:]*indicators[:-1, :-1]
+
+    nobs = len(indicators_joint)
+    corrsum = np.mean(indicators_joint[np.triu_indices(nobs, 1)])
+    return corrsum, indicators_joint
+
+
+def correlation_sums(indicators, max_dim):
+    """
+    Calculate all correlation sums for embedding dimensions 1:max_dim
+
+    Parameters
+    ----------
+    indicators : 2d array
+        matrix of distance threshold indicators
+    max_dim : integer
+        maximum embedding dimension
+
+    Returns
+    -------
+    corrsums : 1d array
+        Correlation sums
+
+    """
+
+    corrsums = np.zeros((1, max_dim))
+
+    corrsums[0, 0], indicators = correlation_sum(indicators, 1)
+    for i in range(1, max_dim):
+        corrsums[0, i], indicators = correlation_sum(indicators, 2)
+
+    return corrsums
+
+
+def _var(indicators, max_dim):
+    """
+    Calculate the variance of a BDS effect
+
+    Parameters
+    ----------
+    indicators : 2d array
+        matrix of distance threshold indicators
+    max_dim : integer
+        maximum embedding dimension
+
+    Returns
+    -------
+    variances : float
+        Variance of BDS effect
+
+    """
+    nobs = len(indicators)
+    corrsum_1dim, _ = correlation_sum(indicators, 1)
+    k = ((indicators.sum(1)**2).sum() - 3*indicators.sum() +
+         2*nobs) / (nobs * (nobs - 1) * (nobs - 2))
+
+    variances = np.zeros((1, max_dim - 1))
+
+    for embedding_dim in range(2, max_dim + 1):
+        tmp = 0
+        for j in range(1, embedding_dim):
+            tmp += (k**(embedding_dim - j))*(corrsum_1dim**(2 * j))
+        variances[0, embedding_dim-2] = 4 * (
+            k**embedding_dim +
+            2 * tmp +
+            ((embedding_dim - 1)**2) * (corrsum_1dim**(2 * embedding_dim)) -
+            (embedding_dim**2) * k * (corrsum_1dim**(2 * embedding_dim - 2)))
+
+    return variances, k
+
+
+def bds(x, max_dim=2, epsilon=None, distance=1.5):
+    """
+    Calculate the BDS test statistic for a time series
+
```

Comment:
also what's the null hypothesis, If pvalue is small, then we reject independence ?